src/transaction_processor.py

```diff
@@ -0,0 +1,182 @@
+        If no entries are present, fallback to get_finalized_block_number().
+        """
+        try:
+            # Query for the maximum block number
+            query_max_block = read_sql_file("src/sql/select_max_block.sql")
+            result = self.db.execute_query(
+                query_max_block, {"chain_name": self.chain_name}
+            )
+            row = result.fetchone()
+            max_block = row[0] if row is not None else None
+
+            if max_block is not None:
+                logger.debug("Fetched max block number from database: %d", max_block)
+
+            # If no entries present, fallback to get_latest_block()
+            if max_block is None:
+                return self.blockchain_data.get_latest_block()
+
+            # Delete entries for the max block from the table
+            delete_sql = read_sql_file("src/sql/delete_entries_max_block.sql")
+            self.db.execute_and_commit(
+                delete_sql, {"chain_name": self.chain_name, "block_number": max_block}
+            )
+            logger.debug("Successfully deleted entries for block number: %s", max_block)
+
+            return max_block
+        except Exception as e:
+            logger.error("Error accessing database: %s", e)
+            return self.blockchain_data.get_latest_block()
+
+    def process(self, start_block: int):
+        """Main Daemon loop that finds imbalances for txs and prices."""
+        previous_block = start_block
+        unprocessed_txs: List[Tuple[str, int, int]] = []
+        logger.info("%s Daemon started. Start block: %d", self.chain_name, start_block)
+
+        while True:
+            try:
+                latest_block = self.blockchain_data.get_latest_block()
+                print(previous_block, latest_block)
+                new_txs = self.blockchain_data.fetch_tx_data(
+                    previous_block, latest_block
+                )
+                all_txs = new_txs + unprocessed_txs
+                unprocessed_txs.clear()
+
+                for tx_hash, auction_id, block_number in all_txs:
+                    try:
+                        self.process_single_transaction(
+                            tx_hash, auction_id, block_number
+                        )
+                    except Exception as e:
+                        unprocessed_txs.append((tx_hash, auction_id, block_number))
+                        logger.error(f"Error processing transaction {tx_hash}: {e}")
+
+                previous_block = latest_block + 1
+                time.sleep(CHAIN_SLEEP_TIME)
+
+            except Exception as e:
+                logger.error(f"Error in processing loop: {e}")
+                time.sleep(CHAIN_SLEEP_TIME)
+
+    def process_single_transaction(
+        self, tx_hash: str, auction_id: int, block_number: int
+    ):
+        """Function processes a single tx to find imbalances, prices."""
+        token_imbalances = self.imbalances.compute_imbalances(tx_hash)
```

Comment:
Instead of making `compute_imbalances` return None, I would raise an error there. That error can then be caught here if desired.

At the moment, the code just raises an error if `token_imbalances is None` so it is equivalent to just raising an error in `compute_imbalances` and not caching it here.